R/predict_spatial.R

```diff
@@ -1,163 +1,73 @@
-    # we need to init the values with a factor class, otherwise setting the
-    # values later on causes conversion troubles
-    target_raster = raster::raster(nrows = raster::nrow(object),
-      ncols = raster::ncol(object),
-      crs = raster::crs(object))
-  }
-  raster::readStart(object)
-  newdata_pred = as.data.table(raster::getValues(object))
-  raster::readStop(object)
-
-  pred = learner$predict_newdata(newdata_pred)
-
-  if (!is.null(filename)) {
-    target_raster = raster::setValues(target_raster, pred$response)
-    raster::writeRaster(target_raster, filename, overwrite = overwrite)
-  }
-  return(pred)
-}
-
-#' @export
-#' @rdname predict_spatial_newdata
-predict_spatial_newdata.sf = function(learner, object, filename = NULL, overwrite = FALSE, quiet = FALSE) {
-
-  if (!is.null(filename)) {
-    assert_path_for_output(filename, overwrite = overwrite)
-  }
-  newdata_pred = as.data.table(object)
-  newdata_pred$geometry = NULL
-  attr(newdata_pred, "sf_column") = NULL
-  attr(newdata_pred, "agr") = NULL
-
-  pred = learner$predict_newdata(newdata_pred)
-
-  if (!is.null(filename)) {
-    sf_pred = sf::st_as_sf(data.frame(pred = pred$response, geometry = object$geometry))
-    sf::st_write(sf_pred, filename, quiet = quiet)
-  }
-  return(pred)
-}
-
+#' stack = demo_stack_spatraster(size = 1)
+#' value = data.table::data.table(ID = c(0, 1), y = c("negative", "positive"))
+#' terra::setCats(stack, layer = "y", value = value)
+#'
+#' # create backend
+#' backend = as_data_backend(stack)
+#' task = as_task_classif(backend, target = "y", positive = "positive")
+#' # train
+#' learner = lrn("classif.featureless")
+#' learner$train(task, row_ids = sample(1:task$nrow, 50))
+#' ras = predict_spatial(task, learner)
+#' ras
 #' @export
-#' @rdname predict_spatial_newdata
-predict_spatial_newdata.stars = function(learner, object, filename = NULL, overwrite = FALSE, quiet = FALSE) {
-
-  if (!is.null(filename)) {
-    assert_path_for_output(filename, overwrite = overwrite)
-  }
-  newdata_pred = as.data.table(split(object, "band"))
-
-  if (any(c("x", "y") %in% colnames(newdata_pred))) { # nocov start
-    if (!quiet) {
-      messagef("Dropping coordinates 'x' and 'y' as they are
-        most likely coordinates. If you want to have these variables included,
-        duplicate them in the stars objects using a different name.
-        To silence this message, set 'quiet = TRUE'.", wrap = TRUE)
-      # nocov end
-    }
-    newdata_pred[, c("x", "y")] = list(NULL)
-  }
-
-  pred = learner$predict_newdata(newdata_pred)
-  # single band object
-  stars_pred = object[, , , 1]
-
-  stars::st_as_stars(dimensions = stars::st_dimensions(object))
-
-  names(stars_pred) = "pred"
-  stars_pred$pred = pred$response
-
-  if (!is.null(filename)) {
-    stars::write_stars(stars_pred, filename, quiet = quiet)
-  }
-  return(pred)
+predict_spatial = function(task, learner, chunksize = 1000L, format = "terra") {
```

Comment:
Maybe use  `chunksize = 100L`? 1000 was way to much for my local machine with parallelization.

Reply:
500? I fear that we might limit too much by default.

When I used 1000L it used about 3 GB memory which I thought is about right. How much was used on your side?